File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2){

    struct ListNode temp;
    struct ListNode* ret = &temp;
    
    while(list1 && list2)
    {
        if(list1->val > list2->val)
        {
            ret->next = list2;
            list2 = list2->next;
        }
        else
        {
            ret->next = list1;
            list1 = list1->next;
        }
        ret = ret->next;
    }
    
    ret->next = list1 ? list1 : list2;
    return temp.next;
    
}

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        def build(r, l1, l2):
            if not l1 or not l2:
                r.next = l1 if l1 else l2
                return
            else:
                if l1.val < l2.val:
                    r.next = l1
                    l1 = l1.next
                else:
                    r.next = l2
                    l2 = l2.next
                build(r.next, l1, l2)
        
        ret = ListNode()
        build(ret, list1, list2)
        return ret.next